geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/data/HBaseFeatureSource.scala

```diff
@@ -0,0 +1,146 @@
+
+  import filter._
+  override def getReaderInternal(query: Query): FR =
+    rewriteFilterInCNF(query.getFilter) match {
+      case a: And            => and(a)
+      case _                 => throw new RuntimeException("Not yet supported")
+    }
+
+
+  override def canFilter: Boolean = true
+  override def canSort: Boolean = true
+
+  private def and(a: And): FR = {
+    // TODO: currently assumes BBOX then DURING
+    import filter._
+
+    val dtFieldName = Some(sft.getDescriptor(dtgIndex).getLocalName)
+    val (i, _) = partitionTemporal(a.getChildren, dtFieldName)
+    val interval = FilterHelper.extractTemporal(dtFieldName)(i)
+
+    val (b, _) = partitionGeom(a, sft)
+    val geom = FilterHelper.extractGeometry(b.head.asInstanceOf[BinarySpatialOperator]).head
+
+    val env = geom.getEnvelopeInternal
+    val (lx, ly, ux, uy) = (env.getMinX, env.getMinY, env.getMaxX, env.getMaxY)
+
+    val epochWeekStart = Weeks.weeksBetween(EPOCH, interval.getStart)
+    val epochWeekEnd = Weeks.weeksBetween(EPOCH, interval.getEnd)
+    val weeks = scala.Range.inclusive(epochWeekStart.getWeeks, epochWeekEnd.getWeeks)
+    val lt = secondsInCurrentWeek(interval.getStart, epochWeekStart)
+    val ut = secondsInCurrentWeek(interval.getEnd, epochWeekStart)
+
+    val kryoFeatureSerializer = new KryoFeatureSerializer(sft)
+    if (weeks.length == 1) {
+      val z3ranges = Z3_CURVE.ranges(lx, ly, ux, uy, lt, ut, 8)
+      // TODO: cache serializers
+      new HBaseFeatureReader(
+        ds.getZ3Table(sft), sft, weeks.head, z3ranges,
+        Z3_CURVE.normLon(lx), Z3_CURVE.normLat(ly), interval.getStart.getMillis,
+        Z3_CURVE.normLon(ux), Z3_CURVE.normLat(uy), interval.getEnd.getMillis,
+        kryoFeatureSerializer)
+    } else {
+      val oneWeekInSeconds = Weeks.ONE.toStandardSeconds.getSeconds
+      val head +: xs :+ last = weeks.toList
+      // TODO: ignoring seconds for now
+      val z3ranges = Z3_CURVE.ranges(lx, ly, ux, uy, 0, oneWeekInSeconds, 8)
+      val middleQPs = xs.map { w =>
+        new HBaseFeatureReader(ds.getZ3Table(sft), sft, w, z3ranges,
+          Z3_CURVE.normLon(lx), Z3_CURVE.normLat(ly), interval.getStart.getMillis,
+          Z3_CURVE.normLon(ux), Z3_CURVE.normLat(uy), interval.getEnd.getMillis,
+          kryoFeatureSerializer)
+      }
+      val sr = new HBaseFeatureReader(ds.getZ3Table(sft), sft, head, z3ranges,
+        Z3_CURVE.normLon(lx), Z3_CURVE.normLat(ly), interval.getStart.getMillis,
+        Z3_CURVE.normLon(ux), Z3_CURVE.normLat(uy), interval.getEnd.getMillis,
+        kryoFeatureSerializer)
+
+      val er = new HBaseFeatureReader(ds.getZ3Table(sft), sft, last, z3ranges,
+        Z3_CURVE.normLon(lx), Z3_CURVE.normLat(ly), interval.getStart.getMillis,
+        Z3_CURVE.normLon(ux), Z3_CURVE.normLat(uy), interval.getEnd.getMillis,
+        kryoFeatureSerializer)
+
+      val readers = Seq(sr) ++ middleQPs ++ Seq(er)
+
+      new FeatureReader[SimpleFeatureType, SimpleFeature] {
```

Comment:
I'd strongly prefer to see the query plan iteration/consumption/joining happening behind the FeatureReader interface.